src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_limit_aggregate.rs

```diff
@@ -21,69 +20,209 @@ use crate::optimizer::rule::TransformResult;
+            let agg_final = s_expr.child(0)?;
+            let agg_partial = agg_final.child(0)?;
+
+            let mut agg_limit: Aggregate = agg_partial.plan().clone().try_into()?;
+
+            let sort_items = agg_limit
+                .group_items
+                .iter()
+                .map(|g| SortItem {
+                    index: g.index,
+                    asc: true,
+                    nulls_first: false,
+                })
+                .collect::<Vec<_>>();
+            agg_limit.rank_limit = Some((sort_items.clone(), count));
+
+            let sort = Sort {
+                items: sort_items.clone(),
+                limit: Some(count),
+                after_exchange: None,
+                pre_projection: None,
+                window_partition: vec![],
+            };
+
+            let agg_partial = SExpr::create_unary(
+                Arc::new(RelOperator::Aggregate(agg_limit)),
+                Arc::new(agg_partial.child(0)?.clone()),
+            );
+            let agg_final = agg_final.replace_children(vec![agg_partial.into()]);
+            let sort = SExpr::create_unary(Arc::new(RelOperator::Sort(sort)), agg_final.into());
+            let mut result = s_expr.replace_children(vec![Arc::new(sort)]);
+
+            result.set_applied_rule(&self.id);
+            state.add_result(result);
+        }
+        Ok(())
+    }
+
+    fn apply_sort(
+        &self,
+        s_expr: &SExpr,
+        state: &mut TransformResult,
+    ) -> databend_common_exception::Result<()> {
+        let sort: Sort = s_expr.plan().clone().try_into()?;
+        let mut has_eval_scalar = false;
+        let agg = match s_expr.child(0)?.plan().rel_op() {
+            RelOp::Aggregate => s_expr.child(0)?,
+            RelOp::EvalScalar => {
+                has_eval_scalar = true;
+                s_expr.child(0)?.child(0)?
+            }
+            _ => return Ok(()),
+        };
+
+        let agg_limit_expr = agg.child(0)?;
+        let mut agg_limit: Aggregate = agg_limit_expr.plan().clone().try_into()?;
+
+        if agg_limit.mode != AggregateMode::Partial {
+            return Ok(());
+        }
+
+        if let Some(limit) = sort.limit {
+            let is_order_subset = sort
+                .items
+                .iter()
+                .all(|k| agg_limit.group_items.iter().any(|g| g.index == k.index));
 
-            agg_limit.limit = Some(agg_limit.limit.map_or(count, |c| cmp::max(c, count)));
-            let agg = SExpr::create_unary(
+            if !is_order_subset {
+                return Ok(());
+            }
+            let mut sort_items = Vec::with_capacity(agg_limit.group_items.len());
+            let mut not_found_sort_items = vec![];
+            for i in 0..agg_limit.group_items.len() {
+                let group_item = &agg_limit.group_items[i];
+                if let Some(sort_item) = sort.items.iter().find(|k| k.index == group_item.index) {
```

Comment:
It looks like we can avoid using `if let Some()`, since `is_order_subset` is true here.

Reply:
eg:
```
select a, b from t group by a,b  order by b desc;
```

`b` is subset of `a,b`, here we inject sort rank `b desc, a asc` for better filter ratio in sort partial.

Reply:
> eg:
> 
> ```
> select a, b from t group by a,b  order by b desc;
> ```
> 
> `b` is subset of `a,b`, here we inject sort rank `b desc, a asc` for better filter ratio in sort partial.

I misunderstood this, thanks for the explanation.